### PR TITLE
[rebranch] Use new AddressSanitizerOptions in pass construction

### PR DESCRIPTION
**Pull request by bnbarham**
llvm/llvm-project `3131714f8daca338492a7d5b189e4d63131bc808` changed
`ModuleAddressSanitizerPass` to take an `AddressSanitizerOptions`
instead. `b7fd30eac3183993806cc218b6deb39eb625c083` also removed
`AddressSanitizerPass` after it was rolled into
`ModuleAddressSanitizerPass` in
`b7fd30eac3183993806cc218b6deb39eb625c083`.

Update IRGen to use the new options and remove the additional pass.

### CODE REVIEW

**Comment by bnbarham** [comment]
@swift-ci please test

File: lib/IRGen/IRGen.cpp [hunk 1]
```diff
@@ -484,21 +484,17 @@ performOptimizationsUsingNewPassManger(const IRGenOptions &Opts,
...
+      Opts.UseAfterReturn = llvm::AsanDetectStackUseAfterReturnMode::Runtime;
       MPM.addPass(
           RequireAnalysisPass<ASanGlobalsMetadataAnalysis, llvm::Module>());
-      MPM.addPass(ModuleAddressSanitizerPass(false, Recover,
-                                             ModuleUseAfterScope,
-                                             UseOdrIndicator, DestructorKind));
-      MPM.addPass(createModuleToFunctionPassAdaptor(AddressSanitizerPass(
-          {false, Recover, UseAfterScope, UseAfterReturn})));
+      MPM.addPass(ModuleAddressSanitizerPass(
+          Opts, /*UseGlobalGC=*/false, Opts.SanitizeAddressUseODRIndicator,
```
Comment (aschwaighofer) [inline lib/IRGen/IRGen.cpp:496]: I think we should set UseGlobalGC to true based on that is what we used to do:

https://github.com/apple/swift/blob/ab4fa0a2046f3d7ba883f6a7d2c9b621d8bedbf3/lib/IRGen/IRGen.cpp#L145

File: lib/IRGen/IRGen.cpp [hunk 2]
```diff
@@ -484,21 +484,17 @@ performOptimizationsUsingNewPassManger(const IRGenOptions &Opts,
...
+      Opts.CompileKernel = false;
+      Opts.Recover = bool(Opts.SanitizersWithRecoveryInstrumentation &
                           SanitizerKind::Address);
-      bool UseAfterScope = false;
-      bool ModuleUseAfterScope = true;
-      bool UseOdrIndicator = Opts.SanitizeAddressUseODRIndicator;
-      llvm::AsanDtorKind DestructorKind = llvm::AsanDtorKind::Global;
-      llvm::AsanDetectStackUseAfterReturnMode UseAfterReturn =
-          llvm::AsanDetectStackUseAfterReturnMode::Runtime;
+      Opts.UseAfterScope = true;
```
Comment (aschwaighofer) [inline lib/IRGen/IRGen.cpp:491]: I think `Opts.UseAfterScope` should be set to `false` to mirror what clang defaults to. It seems to default to false unless you add a flag AFAICT.

**Comment by bnbarham** [comment]
@swift-ci please test

**Review by aschwaighofer — APPROVED** [review]
Thank you

**Comment by bnbarham** [comment]
@swift-ci please test

**Pull request merged**